Add Unifiedpush support

p1gp1g opened the pull request:
### SDK

The SDK exposes the UnifiedPush dependency directly with a 'api' dependency so it can be used to un/register, select the default distributor etc.

The library exposes the UnifiedPush broadcast receiver by itself and forward events to a (non-exported) service the application must implement. That way, we reduce risk of any breaking change.
For that reason, I've hesitated to rewrite PushMessage, PushEndpoint, and FailedReason, to avoid passing classes defined in connector's library.

Any message that haven't been correctly decrypted is ignored. For information, the connector library doesn't do that by itself for backward compatibility, if the application doesn't use web push correctly.

### Client sample

There is a `MockApplicationServer` that emulate a remote application server. Client<->Server calls are emulated with functions of `MockApplicationServer.MockApi`.

I have added some horizontal dividers to add some clarity on the UI.

I've added a new section for push messages:
if unregistered: a button to register
if registered: a button to unregister and a button to send a test notification

The application register to the distributor with VAPID enabled, the MockApplicationServer sends the push message with VAPID.

When receiving a push notification, a notification is displayed with the content of the test message.

khvMX commented:
Hi @p1gp1g. I have questions which I hope you can shed some light on. I'm sorry if I'm asking things which have been clarified already.
1. This PR would embed UnifiedPush directly in "COVESA SDK", without clear separation of its implementation and dependencies from other future "libraries". Have you considered some modularization approach to avoid mixing this with previous and future changes? I'm also thinking about versioning separate libraries within COVESA SDK separately.
2. Should we have different sample apps for different libraries?
3. Is "EventBus" dependency really necessary with current Kotlin and Jetpack state of the art capabilities?

p1gp1g commented:
Hi @khvMX,

> 1. This PR would embed UnifiedPush directly in "COVESA SDK", without clear separation of its implementation and dependencies from other future "libraries". Have you considered some modularization approach to avoid mixing this with previous and future changes? I'm also thinking about versioning separate libraries within COVESA SDK separately.

Indeed, I think there are 2 ways to expose the feature: directly as an api dependency, and the service uses intent with the library classes (https://github.com/COVESA/covesa-aosp-sdk/pull/3/files#diff-dbdd43b7bfea8a29b76c088d80d8d17974ee18ddb299a354f0084cea5f853a49R9-R11) or write new classes that call the library ones. The first solution takes advantage of new features (if any) more easily, the second improves control on the updates.

I asked Maximilian, and he said that one or the other is fine to open the PR, then we can discuss about it.

Let me know if you prefer the 2nd option, it is very fast to do 

> 2. Should we have different sample apps for different libraries?

Both solution make sense. Having a different sample per feature would probably make it clearer. Let me know if you want to split the sample

> 3. Is "EventBus" dependency really necessary with current Kotlin and Jetpack state of the art capabilities?

I've removed it, also it wasn't very clean how the PushUiState was updated, I've fixed that

khvMX commented:
> Indeed, I think there are 2 ways to expose the feature: directly as an api dependency, and the service uses intent with the library classes (https://github.com/COVESA/covesa-aosp-sdk/pull/3/files#diff-dbdd43b7bfea8a29b76c088d80d8d17974ee18ddb299a354f0084cea5f853a49R9-R11) or write new classes that call the library ones. The first solution takes advantage of new features (if any) more easily, the second improves control on the updates.
> 
> I asked Maximilian, and he said that one or the other is fine to open the PR, then we can discuss about it.
> 
> Let me know if you prefer the 2nd option, it is very fast to do

Good point, let's create an issue for that here? It would be great if others would join the discussion. On some of the meetings someone said that "COVESA SDK clients' efforts must be reduced to the absolute minimum", which I support too. Also, reducing API surface would allow for easier under-the-hood changes, albeit for the cost of client flexibility.

khvMX commented:
> > 3\. Should we have different sample apps for different libraries?
> 
> Both solution make sense. Having a different sample per feature would probably make it clearer. Let me know if you want to split the sample

Similar to the above, we probably should create another issue for that, simply to come back to this question once we have at least 3 different SDK features to see any commonalities between them.

p1gp1g commented:
Do you prefer issues or github discussions ? and would you like me to open them?

khvMX commented:
> Do you prefer issues or github discussions ? and would you like me to open them?

Let's move these to (potentially separate) discussions. Could you please create them stating how you have done it for UnifiedPush for now?

As for this PR, it LGTM.

p1gp1g commented:
Discussions for the 2 subjects:
https://github.com/COVESA/covesa-aosp-sdk/discussions/4
https://github.com/COVESA/covesa-aosp-sdk/discussions/5

khvMX commented on api/client/src/main/kotlin/global/covesa/sdk/api/client/push/data/FailedReason.kt:
```diff
@@ -0,0 +1,39 @@
+import org.unifiedpush.android.connector.FailedReason as UFailedReason
+
+/**
+ * A [registration request][TODO()] may fail for different reasons.
```
What does this `TODO()` mean?

khvMX commented on samples/client/build.gradle.kts:
```diff
@@ -17,6 +17,8 @@ dependencies {
     implementation(platform(libs.androidx.compose.bom))
     implementation(libs.androidx.material3)
     implementation(libs.kotlinxCollectionsImmutable)
+    implementation(libs.volley)
```
As long as we are in the same sample app, it would help if we explain what these are for.
Maybe add a comment for each of these dependencies saying what they have been introduced for?

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/MainActivity.kt:
```diff
@@ -6,8 +6,13 @@ import androidx.activity.compose.setContent
 import global.covesa.sdk.client.ui.theme.CovesaSDKTheme
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.onEach
```
unused

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/MainActivity.kt:
```diff
@@ -27,5 +33,14 @@ class MainActivity : ComponentActivity() {
+        subscribeActions()
+    }
+
+    private fun subscribeActions() {
```
Unclear function name. Not even hinting about push notifications.

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/MainActivity.kt:
```diff
@@ -27,5 +33,14 @@ class MainActivity : ComponentActivity() {
+    }
+
+    private fun subscribeActions() {
+        CoroutineScope(Dispatchers.IO).launch {
```
Memory leak? Would this scope be ever destroyed?

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/MainViewModel.kt:
```diff
@@ -27,6 +32,9 @@ class MainViewModel(
     var installedServicesUiState by mutableStateOf(InstalledServicesUiState())
         private set
 
+    var pushUiState by mutableStateOf(context?.let { PushUiState(it) } ?: PushUiState())
```
UiState should not be dependent on `context`. Is it just initial value? You can get value of `PushManager.getAckDistributor(context)` in `Activity` and pass it to the `MainViewModel` constructor instead.

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/push/ActionEvent.kt:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    companion object {
+        private val _events = MutableSharedFlow<ActionEvent>()
```
Why not `MutableStateFlow` instead? Isn't it just for Activity to observe?

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/push/ActionEvent.kt:
```diff
@@ -0,0 +1,41 @@
+        SendNotification,
+    }
+
+    fun handleAction(activity: Activity) {
```
This method should only be visible for `Activity`, so maybe move it there? Also the business logic probably belongs to `Activity`, rather than to some kind of a message queue.

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/push/ActionEvent.kt:
```diff
@@ -0,0 +1,41 @@
+        PushManager.tryUseCurrentOrDefaultDistributor(
+            activity
+        ) {
+            PushManager.register(
```
Aren't you supposed to check the `success` argument before calling `register()`? Also, this API seems redundant. 1 call to `PushManager` should be sufficient (see my `PushManager` comments).

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/push/MockApplicationServer.kt:
```diff
@@ -0,0 +1,221 @@
+
+/**
+ * This class emulates an application server, this should not be implemented
+ * by real applications.
```
It would help if it says what should be done instead.

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/push/MockApplicationServer.kt:
```diff
@@ -0,0 +1,221 @@
+ * This class emulates an application server, this should not be implemented
+ * by real applications.
+ */
+class MockApplicationServer(private val context: Context) {
```
In Android, Mocks usually refer to class with stub methods which do not do anything, I would tend to call this class a `FakeApplicationServer` instead.

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/push/PushServiceImpl.kt:
```diff
@@ -0,0 +1,52 @@
+     * Update the UI
+     */
+    private fun updateRegistrationState(registered: Boolean) {
+        CoroutineScope(Dispatchers.IO).launch {
```
again, memory leak?
you probably need scope injected here

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/ui/MainUi.kt:
```diff
@@ -73,10 +81,51 @@ fun MainUi(viewModel: MainViewModel) {
             LightStates(lightsUiState.lights)
         }
+
+        HorizontalDivider(
```
On a smaller screen, this content does not fit, maybe make the whole screen scrollable with `Scaffold`?

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/ui/PushUiState.kt:
```diff
@@ -0,0 +1,12 @@
+data class PushUiState (
+    val registered: Boolean = false
+) {
+    constructor(context: Context) : this(
```
probably can be injected with a `boolean` instead, see above

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/MainActivity.kt:
```diff
@@ -18,6 +23,7 @@ class MainActivity : ComponentActivity() {
         setContent {
             val viewModel: MainViewModel = viewModel {
                 MainViewModel(
+                    context = this@MainActivity,
```
can be avoided (see other comments)

khvMX commented on samples/client/src/main/java/global/covesa/sdk/client/MainActivity.kt:
```diff
@@ -27,5 +33,14 @@ class MainActivity : ComponentActivity() {
+    private fun subscribeActions() {
+        CoroutineScope(Dispatchers.IO).launch {
+            ActionEvent.events.collect {
+                it.handleAction(this@MainActivity)
```
You are giving a reference to an object (`MainActivity`) which may be destroyed to a global event queue, which may try to use it when it's already destroyed. We need some common scope for these.

khvMX reviewed (changes requested):
Sorry I did not take time before to properly review.
I have left some comments, which are mostly my concerns.
What is also missing is updating the READMEs to reflect the fact that we now have 2 different kinds of services and 2 quite different examples of APIs for apps to use. I will upload some README changes to this branch, so that we can get it reviewed together.

p1gp1g commented:
Thank you @khvMX , most of the comments have been resolved

khvMX commented:
On my emulator the client sample is now crashing with this:
```
2024-12-16 09:10:34.099 12966-12966 AndroidRuntime          global.covesa.sdk.client             E  FATAL EXCEPTION: main
                                                                                                    Process: global.covesa.sdk.client, PID: 12966
                                                                                                    java.lang.IllegalStateException: Vertically scrollable component was measured with an infinity maximum height constraints, which is disallowed. One of the common reasons is nesting layouts like LazyColumn and Column(Modifier.verticalScroll()). If you want to add a header before the list of items please add a header as a separate item() before the main items() inside the LazyColumn scope. There are could be other reasons for this to happen: your ComposeView was added into a LinearLayout with some weight, you applied Modifier.wrapContentSize(unbounded = true) or wrote a custom layout. Please try to remove the source of infinite constraints in the hierarchy above the scrolling container.
```

I believe we're nesting `LazyColumn` in a `Column(Modifier.verticalScroll())`

p1gp1g commented:
Thanks, fixed

khvMX reviewed (approved):
LGTM. We can clean up or change things related to Singleton usage later on.

khvMX commented:
Hey @p1gp1g, I have just attempted to test it with Sunup on a vanilla AAOS emulator. Registration works, but not the noficitation. Is it dependent on the Mozilla PR? This is what I get when I click on "Notify":

```
2025-02-14 12:05:27.244  1665-1715  UnifiedPush             global.covesa.sdk.client             D  Found distributor with package name org.unifiedpush.distributor.sunup
2025-02-14 12:05:27.244  1665-1715  UnifiedPush             global.covesa.sdk.client             D  Found saved distributor.
2025-02-14 12:05:27.244  1665-1715  Compatibil...geReporter global.covesa.sdk.client             D  Compat change id reported: 160794467; UID 1010227; state: DISABLED
2025-02-14 12:05:27.265  1665-1702  EGL_emulation           global.covesa.sdk.client             D  app_time_stats: avg=46.98ms min=7.72ms max=963.66ms count=50
2025-02-14 12:05:28.483  1665-1715  UnifiedPush             global.covesa.sdk.client             D  Found distributor with package name org.unifiedpush.distributor.sunup
2025-02-14 12:05:28.483  1665-1715  UnifiedPush             global.covesa.sdk.client             D  Found saved distributor.
2025-02-14 12:05:28.487  1665-1702  EGL_emulation           global.covesa.sdk.client             D  app_time_stats: avg=1221.58ms min=1221.58ms max=1221.58ms count=1
2025-02-14 12:05:28.672  1665-1720  InternalPu...viceClient global.covesa.sdk.client             I  connecting to remote service
2025-02-14 12:05:29.586  1665-1702  EGL_emulation           global.covesa.sdk.client             D  app_time_stats: avg=99.54ms min=11.59ms max=886.22ms count=11
2025-02-14 12:05:29.676  1665-1717  InternalPu...viceClient global.covesa.sdk.client             D  disconnecting from remote service
2025-02-14 12:05:29.775  1665-1818  Volley                  global.covesa.sdk.client             E  [5739] NetworkUtility.shouldRetryException: Unexpected response code 401 for https://updates.push.services.mozilla.com/wpush/v2/gAAAAABnryN6GWZqX7MGq3Fwm7kuS18KMIrxjn-ZIbMjx7vzvW3U49m1RHdzIGyD8JIbjAc0JwwQh9mubBzr0CPh3z1P-dM8CsQrYSLwqZ8-pI7DbpAfkt5ew5qwFO8T2eolF3PGSFSXuZXaG8_owsxg8V3R4RbuI_5oy2s7LYa46qupQnPiTGc
2025-02-14 12:05:29.946  1665-1818  Volley                  global.covesa.sdk.client             E  [5739] NetworkUtility.shouldRetryException: Unexpected response code 401 for https://updates.push.services.mozilla.com/wpush/v2/gAAAAABnryN6GWZqX7MGq3Fwm7kuS18KMIrxjn-ZIbMjx7vzvW3U49m1RHdzIGyD8JIbjAc0JwwQh9mubBzr0CPh3z1P-dM8CsQrYSLwqZ8-pI7DbpAfkt5ew5qwFO8T2eolF3PGSFSXuZXaG8_owsxg8V3R4RbuI_5oy2s7LYa46qupQnPiTGc
2025-02-14 12:05:29.947  1665-1665  MockApplicationServer   global.covesa.sdk.client             W  An error occurred:
                                                                                                    com.android.volley.AuthFailureError
                                                                                                    	at com.android.volley.toolbox.NetworkUtility.shouldRetryException(NetworkUtility.java:189)
                                                                                                    	at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:145)
                                                                                                    	at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:132)
                                                                                                    	at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:111)
                                                                                                    	at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:90)
```
